### PR TITLE
Document `InferTy` & co.

### PR DESCRIPTION
**Pull request by camelid**
I finally figured out what `TyVid` means! The name is quite opaque, so I
decided to document it and related types.

I don't know that much about `InferTy` & co., but I was able to *infer*
( :) ) from the names and what I know generally about type inference to
add some basic documentation.

### CODE REVIEW

**Comment by camelid** [comment]
FYI, likely will conflict with #79951 (but this adds documentation and the other one just moves stuff, so it shouldn't a big deal for whomever's PR merges second).

File: compiler/rustc_middle/src/ty/sty.rs [hunk 1]
```diff
@@ -1457,18 +1462,40 @@ impl Atom for RegionVid {
...
+///
+/// E.g., if we have an empty array (`[]`), then we create a fresh
+/// type variable for the element type since we won't know until it's
+/// used what the element type is supposed to be.
 #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, TyEncodable, TyDecodable)]
 #[derive(HashStable)]
 pub enum InferTy {
+    /// A type variable.
     TyVar(TyVid),
+    /// An integral type variable.
```
Comment (jyn514) [inline compiler/rustc_middle/src/ty/sty.rs:1475]:
```suggestion
    /// An integral type variable (`{integer}`).
```

File: compiler/rustc_middle/src/ty/sty.rs [hunk 2]
```diff
@@ -1457,18 +1462,40 @@ impl Atom for RegionVid {
...
+    ///
+    /// These are created when the compiler sees an integer literal like
+    /// `1` that could be several different types (`u8`, `i32`, `u32`, etc.).
+    /// We don't know until it's used what type it's supposed to be, so
+    /// we create a fresh type variable.
     IntVar(IntVid),
+    /// A floating-point type variable.
```
Comment (jyn514) [inline compiler/rustc_middle/src/ty/sty.rs:1482]:
```suggestion
    /// A floating-point type variable (`{float}`).
```

**Pull request merged**